google/cloud/bigtable/test_proxy/cbt_test_proxy.cc

```diff
@@ -0,0 +1,398 @@
+  // than the results from the V2 API. In order not to hide unexpected client
+  // behavior, we do not reorder the results. This means that the same family
+  // name or column qualifier may appear multiple times.
+  //
+  // We do, however, combine consecutive cells with the same family or column
+  // qualifier where possible.
+  v2::Family* family = nullptr;
+  v2::Column* column = nullptr;
+
+  for (Cell const& cell : row.cells()) {
+    if (family == nullptr || family->name() != cell.family_name()) {
+      family = new_row.add_families();
+      family->set_name(cell.family_name());
+      column = nullptr;
+    }
+    if (column == nullptr || column->qualifier() != cell.column_qualifier()) {
+      column = family->add_columns();
+      column->set_qualifier(cell.column_qualifier());
+    }
+    v2::Cell* new_cell = column->add_cells();
+    new_cell->set_timestamp_micros(cell.timestamp().count());
+    new_cell->set_value(std::move(cell.value()));
+  }
+  return new_row;
+}
+
+}  // namespace
+
+grpc::Status CbtTestProxy::CreateClient(
+    ::grpc::ServerContext*, testpb::CreateClientRequest const* request,
+    testpb::CreateClientResponse*) {
+  GCP_LOG(DEBUG) << "searching for: " << request->client_id();
+  if (request->client_id().empty() || request->data_target().empty()) {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
+                        "both `client_id` and `data_target` must be provided.");
+  }
+
+  auto options =
+      Options{}
+          .set<DataEndpointOption>(request->data_target())
+          .set<GrpcCredentialOption>(grpc::InsecureChannelCredentials())
+          .set<MaxConnectionRefreshOption>(std::chrono::milliseconds(0))
+          .set<AppProfileIdOption>(request->app_profile_id());
+
+  if (request->has_per_operation_timeout()) {
+    auto duration = absl::ToChronoMilliseconds(absl::Milliseconds(
+        google::protobuf::util::TimeUtil::DurationToMilliseconds(
+            request->per_operation_timeout())));
+    options
+        .set<DataRetryPolicyOption>(
+            DataLimitedTimeRetryPolicy(duration).clone())
+        // TODO(#4926) - set deadlines using a nicer API.
+        .set<google::cloud::internal::GrpcSetupOption>(
+            [duration](grpc::ClientContext& context) {
+              auto deadline = std::chrono::system_clock::now() + duration;
+              if (context.deadline() >= deadline) {
+                context.set_deadline(deadline);
+              }
+            });
+  }
+
+  std::lock_guard<std::mutex> lk(mu_);
+  auto client = connections_.find(request->client_id());
+  if (client != connections_.end()) {
+    return grpc::Status(
+        grpc::StatusCode::ALREADY_EXISTS,
+        absl::StrCat("Client ", request->client_id(), " already exists."));
+  }
+  connections_.insert({request->client_id(), MakeDataConnection(options)});
+  return grpc::Status();
+}
+
+grpc::Status CbtTestProxy::CloseClient(
```

Comment:
How is `Close()` different from `Remove()`?

Reply:
`Close()` is something the java client can do which shuts down its channels or something. And then that java client stops sending requests even when the client API is called.

There is no equivalent in the C++ client anymore. So we just ignore it.

^The code probably deserves a comment explaining as much.

Reply:
Done.